pkgs/build-support/dotnet/build-dotnet-module/default.nix

```diff
@@ -114,23 +111,35 @@ let
 
       _nugetDeps =
         if (nugetDeps != null) then
-          if lib.isDerivation nugetDeps then
+          if lib.isList nugetDeps then
             nugetDeps
           else
-            mkNugetDeps {
-              inherit name;
-              sourceFile = nugetDeps;
-            }
+            assert (lib.isPath nugetDeps);
+            callPackage nugetDeps { fetchNuGet = fetchNupkg; }
         else
-          throw "Defining the `nugetDeps` attribute is required, as to lock the NuGet dependencies. This file can be generated by running the `passthru.fetch-deps` script.";
+          [ ];
```

Comment:
At first glance, this looks like a breaking change, since we no longer handle `isDerivation nugetDeps`?

I guess you've migrated all nixpkgs packages, but should we be trying to maintain a stable API for out of tree usage?

Reply:
That's a good catch, I meant to keep support for derivations there.  There will still be a breaking change in that `passthru.nugetDeps` will now be a list.  It'll work if you pass it in to `buildDotnetModule`, but it's possible it could be used in other ways.

Instead of a list, we could make it an empty derivation by combining the dependencies as propagated inputs.  I think I'll give that a try because I'm seeing things even in-tree like:

```
  buildInputs = [ dotnet-build.nugetDeps ];
```

which seems like it'll be broken.

Reply:
I guess `[ dotnet-build.nugetDeps ]` is actually still working, because nested lists are flattened in `buildInputs`.  I wonder if that works the same way with structured attributes.  I think I'll fix these in nixpkgs, so it's clearly treated as a list, but at least it's not breaking in some cases.

Reply:
I'm adding a test that covers the types allowed in `nugetDeps`.

Reply:
> I'm adding a test that covers the types allowed in `nugetDeps`.

Awesome 👍 

---

I'm wondering, could we flip the logic to take advantage of `toList`? E.g:

```nix
if isPath nugetDeps then
  # ...
else
  toList nugetDeps
```

Since `toList` internally does `v: if isList v then v else [ v ]`.

Reply:
We could, but that would stop if from checking the type of the object when it's not a list.